runtime/src/main/scala/akka/grpc/internal/AkkaDiscoveryNameResolver.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.grpc.internal
+
+import java.net.{ InetAddress, InetSocketAddress, SocketAddress, URI, UnknownHostException }
+
+import akka.discovery.ServiceDiscovery.ResolvedTarget
+import akka.discovery.{ Lookup, ServiceDiscovery }
+import akka.grpc.GrpcClientSettings
+import io.grpc.{ Attributes, EquivalentAddressGroup, NameResolver, Status }
+import io.grpc.NameResolver.Listener
+
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ ExecutionContext, Promise }
+import scala.util.{ Failure, Success }
+
+class AkkaDiscoveryNameResolver(
+    discovery: ServiceDiscovery,
+    defaultPort: Int,
+    authority: String,
+    portName: Option[String],
+    protocol: Option[String],
+    resolveTimeout: FiniteDuration)(implicit val ec: ExecutionContext)
+    extends NameResolver {
+  // TODO
+  override def getServiceAuthority: String = authority
+
+  val listener: Promise[Listener] = Promise()
+
+  override def start(l: Listener): Unit = {
+    listener.trySuccess(l)
+    lookup(l)
+  }
+
+  override def refresh(): Unit =
+    listener.future.onComplete {
+      case Success(l) => lookup(l)
+      case Failure(_) => // We never fail this promise
+    }
+
+  def lookup(listener: Listener): Unit = {
+    discovery.lookup(Lookup(authority, portName, protocol), resolveTimeout).onComplete {
```

Comment:
I think the name of the `authority` field is confusing. Here, for example, it's used as a service name on a service discovery lookup, but on line 28 it's used as an actual `authority`. Instead, we may need to pass a `serviceName` for the lookups and the value of `GrpcSettings#overrideAuthority` (or `serviceName` if unset).

This change here is related to the line `.forTarget("//" + settings.serviceName)` above.

Reply:
I agree this is somewhat confusing, but I'm following the lead of the upstream impl here (https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/DnsNameResolver.java#L174)

Reply:
Makes sense, but that implementation is only valid for DNS `A` lookups (it's a `DnsNameResolver`) and in that case the name and the authority will likely match. `AkkaDiscoveryNameResolver` may use any discovery the user wants to inject. 

And it doesn't matter anyway because we're using:

```
    builder = settings.overrideAuthority.map(builder.overrideAuthority(_)).getOrElse(builder)
```

in `NettyClientUtils`. :-)